Add Parameter Server Example to CI Tests

pbelevich opened the pull request:
Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51037 Add RPC Pipeline Example to CI Tests
* #51016 Add RPC RNN Example to CI Tests
* **#50930 Add Parameter Server Example to CI Tests**

Differential Revision: [D26049857](https://our.internmc.facebook.com/intern/diff/D26049857)

pbelevich committed:
Add Parameter Server Example to CI Tests

[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"




[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"




[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"




[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs




[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/examples/parameter_server_test.py:
```diff
@@ -1,139 +1,252 @@
 import torch.nn as nn
+import torch.nn.functional as F
 from torch import optim
+import torch.distributed.rpc as rpc
+import torch.distributed.autograd as dist_autograd
+from torch.distributed.optim import DistributedOptimizer
+
+from time import perf_counter
+from threading import Lock
 
-from torch.testing._internal.dist_utils import (
-    dist_init,
-    worker_name,
-)
+from torch.testing._internal.dist_utils import dist_init
 from torch.testing._internal.distributed.rpc.rpc_agent_test_fixture import RpcAgentTestFixture
```
shall we sort the import block

pbelevich replied:
addressed

mrshenli commented on torch/testing/_internal/distributed/rpc/examples/parameter_server_test.py:
```diff
@@ -1,139 +1,252 @@
+# remote_method(Foo.bar, rref, arg1, arg2) is equivalent to calling
+# <foo_instance>.bar(arg1, arg2) on the remote node and getting the result
+# back.
+
+
```
unintentional new lines?

pbelevich replied:
removed, literally copy-pasted from the original source

mrshenli commented on torch/testing/_internal/distributed/rpc/examples/parameter_server_test.py:
```diff
@@ -1,139 +1,252 @@
+                cid) != {}
+            opt.step(cid)
+
+    print("Training complete!")
+    print("Getting accuracy....")
```
does it make sense to suppress some outputs as this is not user facing.

pbelevich replied:
But what if someone needs to debug this test, will one need to add it back?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_agent_test_fixture.py:
```diff
@@ -5,6 +5,10 @@
 
 class RpcAgentTestFixture(ABC):
+    @property
+    def name(self):
```
we can use the existing API:

https://github.com/pytorch/pytorch/blob/f0e72e54cc3049e986adc5bff314dd0b649c780b/torch/testing/_internal/dist_utils.py#L194

pbelevich replied:
addressed

mrshenli commented on torch/testing/_internal/distributed/rpc/examples/parameter_server_test.py:
```diff
@@ -1,139 +1,252 @@
+            loss = F.nll_loss(model_output, target)
+            if i % 5 == 0:
+                print(f"Rank {rank} training batch {i} loss {loss.item()}")
+            dist_autograd.backward(cid, [loss])
```
is it possible to fix input and model init values (say by fixing rand seed), and then have some assert on the loss? (e.g., after a few iterations, the loss should drop)

mrshenli reviewed (approved):
LGTM! Added some minor comments.

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs




[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs




[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs




[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs


Differential Revision: [D26049857](https://our.internmc.facebook.com/intern/diff/D26049857)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs


Differential Revision: [D26049857](https://our.internmc.facebook.com/intern/diff/D26049857)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs


Differential Revision: [D26049857](https://our.internmc.facebook.com/intern/diff/D26049857)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs


Differential Revision: [D26049857](https://our.internmc.facebook.com/intern/diff/D26049857)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs


Differential Revision: [D26049857](https://our.internmc.facebook.com/intern/diff/D26049857)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs


Differential Revision: [D26049857](https://our.internmc.facebook.com/intern/diff/D26049857)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Parameter Server Example to CI Tests"

Resolves one item in #46321

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_parameter_server -vs


Differential Revision: [D26049857](https://our.internmc.facebook.com/intern/diff/D26049857)

[ghstack-poisoned]
